File: temoto_nl_interface/actions/find_coordinates/src/get_coordinates_scripts/getcoord_objectmap_generation.cpp
#include <opencv2/opencv.hpp>
#include <vector>
#include <random>
#include <cmath>
#include <nlohmann/json.hpp>

// Using nlohmann/json for JSON parsing
using json = nlohmann::json;

class ObjectMapGenerator {
private:
    // Random number generator
    std::random_device rd;
    std::mt19937 gen;

    // Generate a random color
    cv::Scalar generateRandomColor() {
        return cv::Scalar(
            gen() % 256,  // Blue
            gen() % 256,  // Green
            gen() % 256   // Red
        );
    }

    // Calculate text color based on background brightness
    cv::Scalar getTextColor(const cv::Scalar& backgroundColor) {
        double brightness = (backgroundColor[0] + backgroundColor[1] + backgroundColor[2]) / 3;
        return (brightness < 128) ? cv::Scalar(0, 255, 0) : cv::Scalar(0, 0, 0);
    }

public:
    ObjectMapGenerator() : gen(rd()) {}

    cv::Mat get(const cv::Mat& robot_map, double resolution, 
                const std::pair<double, double>& origin, 
                const json& items_data) {
        // Create a copy of the robot map
        cv::Mat object_map_color = robot_map.clone();

        // Map dimensions
        int map_height = robot_map.rows;
        int map_width = robot_map.cols;

        // Create an overlay image
        cv::Mat overlay = object_map_color.clone();

        // Padding and border properties
        int rectangle_padding = 5;
        int border_thickness = 1;
        cv::Scalar border_color(0, 0, 0);  // Black border

        // Struct to store label information
        struct LabelInfo {
            std::string text;
            cv::Point top_left;
            cv::Point bottom_right;
            cv::Scalar color;
        };
        std::vector<LabelInfo> labels_to_draw;

        // Iterate over items
        for (const auto& [item_class, items_list] : items_data["items"].items()) {
            for (const auto& item : items_list) {
                std::string item_id = item["id"];
                
                // Extract coordinates and dimensions
                double coord_x = item["coordinates"]["x"];
                double coord_y = item["coordinates"]["y"];
                double width = item["dimensions"]["width"];
                double height = item["dimensions"]["height"];

                // Convert world coordinates to map pixel coordinates
                int map_x = static_cast<int>((coord_x - origin.first) / resolution);
                int map_y = map_height - static_cast<int>((coord_y - origin.second) / resolution);

                // Convert dimensions to pixels
                int width_px = static_cast<int>(width / resolution);
                int height_px = static_cast<int>(height / resolution);

                // Calculate rectangle coordinates with padding
                cv::Point top_left(
                    std::max(0, std::min(map_width - 1, map_x - width_px / 2 - rectangle_padding)),
                    std::max(0, std::min(map_height - 1, map_y - height_px / 2 - rectangle_padding))
                );
                cv::Point bottom_right(
                    std::max(0, std::min(map_width - 1, map_x + width_px / 2 + rectangle_padding)),
                    std::max(0, std::min(map_height - 1, map_y + height_px / 2 + rectangle_padding))
                );

                // Generate random color
                cv::Scalar color = generateRandomColor();

                // Draw filled rectangle
                cv2::rectangle(overlay, top_left, bottom_right, color, -1);

                // Draw rectangle border
                cv2::rectangle(overlay, top_left, bottom_right, border_color, border_thickness);

                // Draw "X" lines inside the rectangle
                cv2::line(overlay, top_left, bottom_right, border_color, 1);
                cv2::line(overlay, 
                    cv::Point(top_left.x, bottom_right.y), 
                    cv::Point(bottom_right.x, top_left.y), 
                    border_color, 1);

                // Store label information
                labels_to_draw.push_back({item_id, top_left, bottom_right, color});
            }
        }

        // Draw labels
        for (const auto& label : labels_to_draw) {
            int font = cv2::FONT_HERSHEY_SIMPLEX;
            double font_scale = 0.3;
            int thickness_text = 0;

            // Calculate text size
            int baseline;
            cv::Size text_size = cv2::getTextSize(label.text, font, font_scale, thickness_text, &baseline);

            // Calculate text position
            cv::Point text_pos(
                (label.top_left.x + label.bottom_right.x) / 2 - text_size.width / 2,
                (label.top_left.y + label.bottom_right.y) / 2 + text_size.height / 2
            );

            // Determine text and outline colors
            cv::Scalar text_color = getTextColor(label.color);
            cv::Scalar outline_color = (text_color == cv::Scalar(0, 255, 0)) ? cv::Scalar(0, 0, 0) : cv::Scalar(255, 255, 255);

            // Draw text outline
            cv2::putText(overlay, label.text, text_pos, font, font_scale, outline_color, 3, cv2::LINE_AA);

            // Draw text
            cv2::putText(overlay, label.text, text_pos, font, font_scale, text_color, 1, cv2::LINE_AA);
        }

        // Apply transparency
        double alpha = 1.0;
        cv2::addWeighted(overlay, alpha, object_map_color, 1 - alpha, 0, object_map_color);

        return object_map_color;
    }
};

File: temoto_nl_interface/actions/find_coordinates/src/getcoord_py/getcoord_objectmap_generation.py
# getcoord_objectmap_generation.py

import cv2
import numpy as np
import yaml
import random
import math

def get(robot_map, resolution, origin, items_data):
    
    # Map origin and resolution
    origin_x, origin_y = origin[0], origin[1]

    # Use the robot_map as the base image
    object_map_color = robot_map.copy()

    # Get map dimensions
    map_height, map_width, _ = object_map_color.shape

    # Create an overlay image for drawing the rectangles and lines
    overlay = object_map_color.copy()

    # Define padding for the rectangles (in pixels)
    rectangle_padding = 5  # Adjust as needed

    # Define border properties
    border_thickness = 1  # Thickness of the rectangle border in pixels
    border_color = (0, 0, 0)  # Black border

    # List to store label information to draw after all rectangles
    labels_to_draw = []

    # Iterate over each class and their items
    for item_class, items_list in items_data['items'].items():
        for item in items_list:
            item_id = item['id']
            coordinates = item['coordinates']
            dimensions = item['dimensions']

            # Convert world coordinates to map pixel coordinates
            map_x = int((coordinates['x'] - origin_x) / resolution)
            map_y = int((coordinates['y'] - origin_y) / resolution)

            # Adjust for image coordinate system (origin at top-left)
            map_y = map_height - map_y

            # Convert dimensions from meters to pixels
            width_px = int(dimensions['width'] / resolution)
            height_px = int(dimensions['height'] / resolution)

            # Define rectangle coordinates with padding
            top_left = (
                map_x - width_px // 2 - rectangle_padding,
                map_y - height_px // 2 - rectangle_padding
            )
            bottom_right = (
                map_x + width_px // 2 + rectangle_padding,
                map_y + height_px // 2 + rectangle_padding
            )

            # Ensure coordinates are within map bounds
            top_left = (
                max(0, min(map_width - 1, top_left[0])),
                max(0, min(map_height - 1, top_left[1]))
            )
            bottom_right = (
                max(0, min(map_width - 1, bottom_right[0])),
                max(0, min(map_height - 1, bottom_right[1]))
            )

            # Generate a random color for each item
            color = (
                random.randint(0, 255),  # Blue channel
                random.randint(0, 255),  # Green channel
                random.randint(0, 255)   # Red channel
            )

            # Draw the filled rectangle on the overlay image with the random color
            cv2.rectangle(overlay, top_left, bottom_right, color=color, thickness=-1)

            # Draw the rectangle border (outline) on the overlay
            cv2.rectangle(
                overlay,
                top_left,
                bottom_right,
                color=border_color,
                thickness=border_thickness
            )

            # Draw "X" lines inside the rectangle
            cv2.line(
                overlay,
                top_left,
                bottom_right,
                color=border_color,
                thickness=1  # Thin line
            )
            cv2.line(
                overlay,
                (top_left[0], bottom_right[1]),
                (bottom_right[0], top_left[1]),
                color=border_color,
                thickness=1  # Thin line
            )

            # Collect label information for later drawing
            labels_to_draw.append({
                'text': item_id,
                'top_left': top_left,
                'bottom_right': bottom_right,
                'color': color
            })

    # Now, draw all labels to ensure they are on top
    for label in labels_to_draw:
        item_id = label['text']
        top_left = label['top_left']
        bottom_right = label['bottom_right']
        color = label['color']

        # Define padding for the text inside the rectangle
        text_padding = 1  # pixels

        # Choose font and scale
        font = cv2.FONT_HERSHEY_SIMPLEX
        font_scale = 0.3  # Adjusted for better readability
        thickness_text = 0  # Fixed thickness for text

        # Calculate size of the text
        (text_width, text_height), baseline = cv2.getTextSize(item_id, font, font_scale, thickness_text)

        # Calculate center position of the rectangle
        rect_center_x = (top_left[0] + bottom_right[0]) // 2
        rect_center_y = (top_left[1] + bottom_right[1]) // 2

        # Calculate text position to center it
        text_x = rect_center_x - (text_width // 2)
        text_y = rect_center_y + (text_height // 2)

        text_position = (text_x, text_y)

        # Text color should contrast with rectangle color
        # Calculate brightness of the rectangle color
        brightness = (color[0] + color[1] + color[2]) / 3
        if brightness < 128*100:
            text_color = (0, 255, 0)  # White for dark rectangles
            outline_color = (0, 0, 0)     # Black outline for text
        else:
            text_color = (0, 0, 0)        # Black for light rectangles
            outline_color = (255, 255, 255)  # White outline for text

        # Draw text outline for better visibility
        cv2.putText(
            overlay,
            item_id,
            text_position,
            font,
            font_scale,
            outline_color,
            thickness=3,  # Thicker outline
            lineType=cv2.LINE_AA
        )

        # Draw the actual text
        cv2.putText(
            overlay,
            item_id,
            text_position,
            font,
            font_scale,
            text_color,
            thickness=1,
            lineType=cv2.LINE_AA
        )

    # Apply transparency to the overlay
    alpha = 1  # Transparency factor (0.0 - 1.0)
    cv2.addWeighted(overlay, alpha, object_map_color, 1 - alpha, 0, object_map_color)

    # Return the modified object map
    return object_map_color
